chainer/functions/normalization/batch_normalization.py

```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
-            mean = x.mean(axis=axis)
-            var = x.var(axis=axis)
-            var += self.eps
-            self.mean = mean
-            self.var = var
+        if self.train:
+            if self.running_mean is None:
+                self.running_mean = xp.zeros_like(gamma)
+                self.running_var = xp.zeros_like(gamma)
+            else:
+                self.running_mean = xp.array(self.running_mean)
+                self.running_var = xp.array(self.running_var)
+        elif len(inputs) == 5:
+                self.fixed_mean = xp.array(inputs[3])
+                self.fixed_var = xp.array(inputs[4])
 
-        self.std = xp.sqrt(var, dtype=var.dtype)
+        # cuDNN only supports these tensor dimensions because they are
+        # the most commonly used. If there is a need to support other
+        # dimensions with cuDNN, we could consider reshaping the input
+        # into a 2-dim array with channels as second dim and m=<product
+        # of all dimensions except the 2nd dimension> as the first
+        # dimension.
+        self.cudnn_dim_ok = x.ndim == 2 or x.ndim == 4
 
+        cudnn_updated_running_stats = False
         if xp is numpy:
+            if self.train:
+                axis = (0,) + tuple(range(head_ndim, x.ndim))
+                mean = x.mean(axis=axis)
+                var = x.var(axis=axis)
+                var += self.eps
+            else:
+                mean = self.fixed_mean
+                var = self.fixed_var
+            self.std = xp.sqrt(var, dtype=var.dtype)
             x_mu = x - mean[expander]
             self.x_hat = x_mu / self.std[expander]
             y = gamma * self.x_hat
             y += beta
+            # cuDNN Batch normalization requires v5 or greater.
+        elif cuda.cudnn_enabled and self.use_cudnn and self.cudnn_dim_ok and \
+                (_cudnn_version >= 5000):
+            if x.ndim == 4:
+                # for convolutional layer
+                self.mode = libcudnn.CUDNN_BATCHNORM_SPATIAL
+            else:
+                # for linear layer
+                self.mode = libcudnn.CUDNN_BATCHNORM_PER_ACTIVATION
+
+            x = cuda.cupy.ascontiguousarray(x)
+            gamma = cuda.cupy.ascontiguousarray(gamma)
+            beta = cuda.cupy.ascontiguousarray(beta)
+            dtype = x.dtype
+            handle = cudnn.get_handle()
+            x_desc = cudnn.create_tensor_descriptor(_as4darray(x))
+            derivedBnDesc = cudnn.create_uninitialized_tensor_descriptor()
+            libcudnn.deriveBNTensorDescriptor(derivedBnDesc.value,
+                                              x_desc.value, self.mode)
+            one = numpy.array(1, dtype=dtype).ctypes
+            zero = numpy.array(0, dtype=dtype).ctypes
+            y = cuda.cupy.empty_like(x)
+            # Factor used in the moving average
+            factor = 1 - self.decay
+            # computation runningMean
+            if self.mean_cache is None:
```

Comment:
How about moving this block into `if self.train` block?
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R152